src/rgw/rgw_notify.cc

```diff
@@ -2,17 +2,530 @@
+        } else {
+          ldout(cct, 20) << "INFO: removed entries up to: " << end_marker <<  " from queue: " 
+            << queue_name << dendl;
+        }
+      }
+      // TODO: cleanup expired reservations
+    }
+
+sched_next:
+    if (is_idle) {
+      boost::asio::deadline_timer timer(io_context);
+      timer.expires_from_now(boost::posix_time::microseconds(queue_idle_sleep_us));
+	    ldout(cct, 20) << "INFO: queue " << queue_name << " is idle, next processing will happen in: " << queue_idle_sleep_us << " usec" << dendl;
+      boost::system::error_code ec;
+	    timer.async_wait(yield[ec]);
+      // TODO check ec
+	    spawn::spawn(io_context, [this, &queue_name](spawn::yield_context yield) {
+          process_queue(yield, queue_name);
+          });
+    } else {
+	    ldout(cct, 20) << "INFO: queue " << queue_name << " is will be processed again. pending work for this queue: " << pending_work << dendl;
+	    spawn::spawn(io_context, [this, &queue_name](spawn::yield_context yield) {
+        process_queue(yield, queue_name);
+        });
+    }
+  }
+
+  // process all queues
+  // find which of the queues is owned by this daemon and process it
+  void process_queues(spawn::yield_context yield) {
+    auto has_error = false;
+    auto ret = try_to_create_queue_list();
+    if (ret < 0) {
+      // failed to create the queue list object
+      has_error = true;
+      goto sched_next;
+    }
+    {
+      queues_t queues;
+      ret = read_queue_list(queues);
+      if (ret < 0) {
+        // failed to create the queue list object
+        has_error = true;
+        goto sched_next;
+      }
+
+      for (const auto& queue_name : queues.list) {
+        // try to lock the queue to check if it is owned by this rgw
+        // or if ownershif needs to be taken
+        ret = rados::cls::lock::lock(&rados_ioctx, queue_name, queue_name+"_lock", 
+              LOCK_EXCLUSIVE,
+              lock_cookie, 
+              "" /*no tag*/,
+              "" /*no description*/,
+              failover_time,
+              LOCK_FLAG_MAY_RENEW);
+
+        if (ret == -EBUSY) {
+          // lock is already taken by another RGW
+          ldout(cct, 20) << "INFO: queue: " << queue_name << " owned (locked) by another daemon" << dendl;
+          continue;
+        }
+        if (ret < 0) {
+          // failed to lock for another reason, continue to process other queues
+          ldout(cct, 1) << "ERROR: failed renew lock on queue: " << queue_name << ". error: " << ret << dendl;
+          has_error = true;
+        }
+        // add queue to list of owned queues
+        if (owned_queues.insert(queue_name).second) {
+          ldout(cct, 20) << "INFO: queue: " << queue_name << " now owned (locked) by this daemon" << dendl;
+          // start processing this queue
+          spawn::spawn(io_context, [this, &queue_name](spawn::yield_context yield) {
+              process_queue(yield, queue_name);
```

Comment:
it looks like this should capture `queue_name` by value. this lambda is called asynchronously, and `queues_t queues` may have destructed by then